src/compiler/scala/tools/nsc/transform/Mixin.scala

```diff
@@ -266,11 +261,23 @@ abstract class Mixin extends InfoTransform with ast.TreeDSL {
               //     level.
 
               lazy val mixinSuperInterfaces = mixinClass.ancestors.filter(_.isTraitOrInterface)
-              val needsForwarder = competingMethods.exists(m => {
-                !m.owner.isTraitOrInterface ||
-                  (!m.isDeferred && !mixinSuperInterfaces.contains(m.owner))
-              })
-              if (needsForwarder)
+
+              @tailrec
+              def existsCompetingMethod(baseClasses: List[Symbol]): Boolean = baseClasses match {
+                case baseClass :: rest =>
+                  if (baseClass ne mixinClass) {
+                    val m = member.overriddenSymbol(baseClass)
+                    val isCompeting = m.exists && {
+                      !m.owner.isTraitOrInterface ||
+                        (!m.isDeferred && !mixinSuperInterfaces.contains(m.owner))
```

Comment:
How did you weigh the choice between this formulation `mixinSuperInterfaces.contains(m.owner)` and my proposal of `mixinClass.isNonBottomSubclass(m.owner)`? The latter amounts to a binary search over an existing data structure (the base type seq).